util/chunk/chunk.go

```diff
@@ -112,12 +157,63 @@ func (c *Chunk) SetNumVirtualRows(numVirtualRows int) {
 // Reset resets the chunk, so the memory it allocated can be reused.
 // Make sure all the data in the chunk is not used anymore before you reuse this chunk.
 func (c *Chunk) Reset() {
-	for _, c := range c.columns {
-		c.reset()
+	c.GrowAndReset(c.capacity)
+}
+
+// GrowAndReset resets the chunk and grow capacity if need.
+// TODO: this method will be used in continue PR.
+func (c *Chunk) GrowAndReset(capLimit int) {
+	if c.columns == nil {
+		return
+	}
+	needScaleUp, newCap := checkCapacity(c, capLimit, false)
+	if !needScaleUp {
+		for _, col := range c.columns {
+			col.reset()
+		}
+		c.numVirtualRows = 0
+		return
 	}
+	c.capacity = newCap
+	c.columns = resizeColumns(c.columns, newCap)
 	c.numVirtualRows = 0
 }
 
+// checkCapacity checks and return suitable capacity in next execution.
+//
+//    the suitable capacity is determined by this table:
+
+//      condition                            needScaleUp      newCapacity
+//      -------------------------------------------------------------
+//      chk size is empty                       No           keep old cap
+//      chk cap is over limit                   No           keep old cap(=capLimit)
+//      chk size is full                        Yes          old cap x 2(not over capLimit)
+//      chk size is not full & !tryScaleDown    No           keep old cap
+//      chk size is not full & tryScaleDown     No           power of 2 for old chk size
+//
+//    tryScaleDown is an optimization for Renew() situation in which we will reallocate memory in every loop,
+//    and check `chk.NumRows == 0` to stop loop, so for last loop we can use a small chunk to check.
+//    but for Reset() situation in which the chunk keep reusing reallocate a smaller chunk is expensive in most situation.
+//
+func checkCapacity(c *Chunk, capLimit int, tryScaleDown bool) (needScaleUp bool, newCap int) {
```

Comment:
Can we define different `calculateCapacity` functions for `Renew` and `Reset`?
I think that would be easier to read.